### DISCUSSION TITLE
DB_UPDATE_FREQUENCY

### DISCUSSION THREAD

**Discussion by Pythonov** [Kong/kong#9044] [Help]
Hi!
Could you help me?
I have 2 nodes of kong. I created service on 1st node, and then checked services on both nodes
On both nodes it was created at the same time
I repeated that for several times
I don't understand why there is no delay (5 seconds maximum)

**Comment by ms2008** [comment]
Perhaps the following diagram can be a good explanation.

```
   5s
|_____|_____|_____|_____|_____|_____|  ==> node-1 
        |
        At this moment, you created a service on node-1. There only be 1s need to wait for node-2 starting the next poll.
___|_____|_____|_____|_____|_____|___  ==> node-2
```

**Comment by Pythonov** [comment]
Thank you for the answer!
I understand how it works. I've created small script that creates service on 1st node and after that gets all services from both nodes.
Here is code

       Check number of services on both nodes
        node_one = requests.get(
            url='node_1_url:8011/services/',
        )
        print(
          len(node_one['data'])
        )
        node_two = requests.get(
            url='node_2_url:8001/services/',
        )
        print(
          len(node_two['data'])
        )
        
              Create service on 1st node
        requests.post(
            url='node_1_url/services/',
            data={
                "id": sguid,
                "some": "data",
            }
        )
        
              Check number of services on both nodes again
        node_one = requests.get(
            url='node_1_url:8011/services/',
        )
        print(
          len(node_one['data'])
        )
        
        node_two = requests.get(
            url='node_2_url:8001/services/',
        )
        print(
          len(node_two['data'])
        )
        
        requests.delete(
            url='node_1_url/services/guid'
        )


I always get output like that:
7
7
8
8

I've tested it many times and the result is the same.
What am i missing?
I don't believe that i'm so lucky that i always get in THAT moment
```
   5s
|_____|_____|_____|_____|_____|_____|  ==> node-1 
 |
        
___|_____|_____|_____|_____|_____|___  ==> node-2
```

**Reply by flrgh** [reply]
Hi @Pythonov :wave: 

The Kong admin API queries the database directly when invoked. This is why you see the service reported from both nodes instantly (I'm assuming they are in a cluster/using the same DB).

In traditional mode, the `db_update_frequency` setting affects how often Kong checks for cluster level events, such as cache invalidations.
This is closely related to the [`worker_state_update_frequency`](https://docs.konghq.com/gateway/latest/reference/configuration/#worker_state_update_frequency) setting, which affects how often the proxy layer will attempt to rebuild cached objects (such as the router and plugin iterator) if they have been invalidated by a cluster event. Here's some rough pseudo code to illustrate:

```
current_router_version = 1
update_router = false

while true
  sleep($db_update_frequency)
  for event in get_cluster_events()
    if event.type == "invalidation" and event.key == "router:version" then
      update_router = true

# meanwhile, in another thread
while true
  sleep($worker_state_update_frequency)
  if update_router then
    new_router_version = get_router_version_from_db()
    if new_router_version != current_router_version then
      rebuild_router(get_all_routes_from_db())
      update_router = false
      current_router_version = new_router_version
```

If [`worker_consistency`](https://docs.konghq.com/gateway/latest/reference/configuration/#worker_consistency) is set to `strict`, things are slightly different, but I'll leave that out for now as to not complicate things further.

So, generally speaking, if you make changes via the admin API on one node, it will be visible on other nodes right away via the admin API. It is at the proxy layer where the update frequency/interval values make a difference. This is why there is an observable delay between adding a route on one node and the route becoming active/usable on other nodes.
